contracts/oracles/common/CappedOracle.sol

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.25;
+
+import { OracleInterface } from "../../interfaces/OracleInterface.sol";
+import { ensureNonzeroAddress, ensureNonzeroValue } from "@venusprotocol/solidity-utilities/contracts/validators.sol";
+import { IERC20Metadata } from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import { CappedOracle } from "./CappedOracle.sol";
+import { Transient } from "../../lib/Transient.sol";
+
+/**
+ * @title CappedOracle
+ * @notice This oracle fetches the price of a correlated token and caps the growth rate
+ */
+abstract contract CappedOracle is OracleInterface {
+    using Transient for bytes32;
+
+    /// Slot to cache the asset's price, used for transient storage
+    bytes32 public constant CACHE_SLOT = keccak256(abi.encode("venus-protocol/oracle/common/CappedOracle/cache"));
+
+    //// @notice Growth rate percentage in seconds. Ex: 1e18 is 100%
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    uint256 public immutable GROWTH_RATE_PER_SECOND;
+
+    /// @notice Snapshot update interval
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    uint256 public immutable SNAPSHOT_INTERVAL;
+
+    /// @notice Last stored snapshot price
+    uint256 public snapshotPrice;
+
+    /// @notice Last stored snapshot timestamp
+    uint256 public snapshotTimestamp;
+
+    /// @notice Emitted when the snapshot is updated
+    event SnapshotUpdated(uint256 price, uint256 timestamp);
+
+    /// @notice Constructor for the implementation contract.
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(uint256 annualGrowthRate, uint256 snapshotInterval) {
+        SNAPSHOT_INTERVAL = snapshotInterval;
+        GROWTH_RATE_PER_SECOND = (annualGrowthRate) / (365 * 24 * 60 * 60);
+    }
+
+    /**
+     * @notice Returns if the price is capped
+     * @return isCapped Boolean indicating if the price is capped
+     */
+    function isCapped() external view virtual returns (bool) {
+        uint256 price = getUncappedPrice(token());
+
+        uint256 maxAllowedPrice = _getMaxAllowedPrice();
+
+        return ((price > maxAllowedPrice) && (maxAllowedPrice != 0)) ? true : false;
+    }
+
+    /**
+     * @notice Updates the snapshot price and timestamp
+     */
+    function updateSnapshot() public {
+        address asset = token();
+        if (Transient.readCachedPrice(CACHE_SLOT, asset) != 0) {
+            return;
+        }
+        if (block.timestamp - snapshotTimestamp < SNAPSHOT_INTERVAL || SNAPSHOT_INTERVAL == 0) return;
+
+        snapshotPrice = getPrice(asset);
+        snapshotTimestamp = block.timestamp;
+        Transient.cachePrice(CACHE_SLOT, asset, snapshotPrice);
+        emit SnapshotUpdated(snapshotPrice, snapshotTimestamp);
+    }
+
+    /**
+     * @notice Fetches the price of the token
+     * @param asset Address of the token
+     * @return price The price of the token in scaled decimal places
```

Comment:
I would specifically mention that the returned value can be capped to a maximum value taking into account the growth rate